llvm/unittests/Telemetry/TelemetryTest.cpp

```diff
@@ -0,0 +1,336 @@
+};
+
+namespace vendor {
+struct VendorConfig : public Config {
+  VendorConfig(bool Enable) : Config(Enable) {}
+  std::string makeSessionId() override {
+    static int seed = 0;
+    return std::to_string(seed++);
+  }
+};
+
+std::shared_ptr<Config> GetTelemetryConfig(const TestContext &Ctxt) {
+  return std::make_shared<VendorConfig>(/*EnableTelemetry*/ true);
+}
+
+class JsonStorageDestination : public Destination {
+public:
+  JsonStorageDestination(TestContext *Ctxt) : CurrentContext(Ctxt) {}
+
+  Error receiveEntry(const TelemetryInfo *Entry) override {
+    if (Error Err = serializer.init())
+      return Err;
+
+    Entry->serialize(serializer);
+    if (Error Err = serializer.finalize())
+      return Err;
+
+    json::Object copied = *serializer.getOutputObject();
+    CurrentContext->EmittedJsons.push_back(std::move(copied));
+
+    return Error::success();
+  }
+
+  llvm::StringLiteral name() const override { return "JsonDestination"; }
+
+private:
+  TestContext *CurrentContext;
+  JsonSerializer serializer;
+};
+
+struct StartupInfo : public TelemetryInfo {
+  std::string ToolName;
+
+  void serialize(Serializer &serializer) const override {
+    TelemetryInfo::serialize(serializer);
+    serializer.write("ToolName", ToolName);
+  }
+};
+
+struct ExitInfo : public TelemetryInfo {
+  int ExitCode;
+  std::string ExitDesc;
+  void serialize(Serializer &serializer) const override {
+    TelemetryInfo::serialize(serializer);
+    serializer.write("ExitCode", ExitCode);
+    serializer.write("ExitDesc", ExitDesc);
+  }
+};
+
+class TestManager : public Manager {
+public:
+  static std::unique_ptr<TestManager>
+  createInstance(Config *Config, TestContext *CurrentContext) {
+    if (!Config->EnableTelemetry)
+      return nullptr;
+    CurrentContext->ExpectedUuid = Config->makeSessionId();
+    std::unique_ptr<TestManager> Ret = std::make_unique<TestManager>(
+        CurrentContext, CurrentContext->ExpectedUuid);
+
+    // Add a few destinations.
+    Ret->addDestination(
+        std::make_unique<JsonStorageDestination>(CurrentContext));
+
+    return Ret;
+  }
+
+  TestManager(TestContext *Ctxt, std::string Id)
+      : CurrentContext(Ctxt), SessionId(Id) {}
+
+  Error dispatch(TelemetryInfo *Entry) override {
+    Entry->SessionId = SessionId;
+    Error AllErrs = Error::success();
+    for (auto &Dest : Destinations) {
+      if (Error Err = Dest->receiveEntry(Entry))
+        AllErrs = joinErrors(std::move(AllErrs), std::move(Err));
+    }
+    return AllErrs;
+  }
+
+  void addDestination(std::unique_ptr<Destination> Dest) override {
+    Destinations.push_back(std::move(Dest));
+  }
+
+  std::string getSessionId() { return SessionId; }
+
+  Error atStartup(StartupInfo *Info) { return dispatch(Info); }
+
+  Error atExit(ExitInfo *Info) { return dispatch(Info); }
+
+private:
+  TestContext *CurrentContext;
+  const std::string SessionId;
+  std::vector<std::unique_ptr<Destination>> Destinations;
```

Comment:
> It's probably been discussed before, but given the length of this review, I've forgotten what the reasoning was. Could you walk me through why `Destinations` isn't a member of the base class (along with the concrete `addDestination`)?

Are you saying to make `Destination` a class inside `Manager` class? What difference would that make? (other than some name-nesting)

Or are you asking why we need a separate `Destination` class? If so, it's for modularity and/or  code sharing (eg., in LLDB, we would have one common Manager doing most of the stuff, then vendors can supply their own Destinations without having to extend or redefine the shared lldb::Manager)



> To elaborate more on that second point, I'm thinking that `dispatch` could become a concrete method in the base class that calls a virtual `dispatchImp` ...

This doesn't work for LLDB because your proposal assumes the same `dispatchImpl` can be used for all scenarios, which is NOT the case for LLDB. For LLDB, would instead define multiple instrumentation methods for different scenarios. These would collect different data, then eventually delegate to the `dispatch()` method to send the entries off.

Secondly, your design makes it look like we're sending data to two differnt places (one via the `dispatchImpl`, then second to the destinations).

Reply:
> Are you saying to make `Destination` a class inside `Manager` class? What difference would that make? (other than some name-nesting)
> 
> Or are you asking why we need a separate `Destination` class? If so, it's for modularity and/or code sharing (eg., in LLDB, we would have one common Manager doing most of the stuff, then vendors can supply their own Destinations without having to extend or redefine the shared lldb::Manager)

I think you misread my comment. I specifically was referring to the `Destinations` member, i.e. the thing storing the added destinations. So I'm proposing that the `addDestination` method would be part of the base `Manager` and that class would store the added `Destinations`.

Regarding my suggestion with `dispatchImpl`, I was suggesting this could be a hook that downstream vendors (or even tool managers) could implement to attach common data to every dispatched telemetry info. For example, in our downstream implementation, all events receive certain properties of the application and host. This is independent of what you're talking about with LLDB's additional entry points. Anyway, the `dispatchImpl` aspect is mostly tangential to the core of what I was suggesting, namely that the generic dispatch behaviour of iterating over `Destinations` and sending the event to them should be in the base class. I don't think this impacts how LLDB would do things.

Reply:
K, how about making the `dispatch()` a virtual but with a default implementation. (we might want to override it to dispatch to different destinations in parallel or whatever fashion ...)

Then,  call the `dispatchImpl` something like `preDispatch`? (since it's invoked prior to the entry being sent to the Destination list)

Reply:
Makes sense to me. I'm also happy if you just want to leave off `preDispatch` until we have a use-case for it.